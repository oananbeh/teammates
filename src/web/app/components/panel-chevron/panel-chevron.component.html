<button type="button" class="chevron btn-course btn btn-primary btn-sm" aria-label="Expand panel" [attr.aria-expanded]=isExpanded>
  <i class="fas fa-chevron-down" [style.color]="chevronColor" *ngIf="!isExpanded"></i>
  <i class="fas fa-chevron-up" [style.color]="chevronColor" *ngIf="isExpanded"></i>
</button>
