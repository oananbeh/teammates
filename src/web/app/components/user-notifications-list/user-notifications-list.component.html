<div id="notifications-timezone">
  <p>All dates are displayed in {{ timezone }} time.</p>
</div>
<div class="row mb-3" *ngIf="notificationTabs.length > 0">
  <div class="col-12">
    <div class="float-end">
      <strong class="d-inline"> Sort By: </strong>
      <div class="btn-group" data-toggle="buttons">
        <button id="sort-end-time" class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.NOTIFICATION_END_TIME)" (click)="sortNotificationsBy(SortBy.NOTIFICATION_END_TIME)">End Time</button>
        <button id="sort-start-time" class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.NOTIFICATION_START_TIME)" (click)="sortNotificationsBy(SortBy.NOTIFICATION_START_TIME)">Start Time</button>
      </div>
    </div>
  </div>
</div>

<tm-loading-retry [shouldShowRetry]="hasLoadingFailed" [message]="'Failed to load notifications'" (retryEvent)="loadNotifications()">
  <div *tmIsLoading="isLoadingNotifications">
    <div class="alert alert-warning margin-top-30px" role="alert" *ngIf="notificationTabs.length === 0 && !isLoadingNotifications">
      There are no active notifications.
    </div>
  </div>
  <ng-container>
    <div id="notification-tabs">
      <div *ngFor="let notificationTab of notificationTabs">
        <div [id]="notificationTab.notification.notificationId" class="card">
          <div class="card-header cursor-pointer mb-0" [ngClass]="notificationTab.notification.style | notificationStyleClass" (click)="toggleCard(notificationTab)">
            <strong class="text-break">{{notificationTab.notification.title}} [{{notificationTab.startDate}} - {{notificationTab.endDate}}]</strong>
            <div class="card-header-btn-toolbar">
              <tm-panel-chevron [isExpanded]="notificationTab.hasTabExpanded"></tm-panel-chevron>
            </div>
          </div>

          <div class="mb-0" *ngIf="notificationTab.hasTabExpanded" @collapseAnim>
            <div [ngClass]="getBodyTextClass(notificationTab)">
              <div class="notification-message" [innerHtml]="notificationTab.notification.message"></div>
            </div>
            <div class="d-flex flex-row-reverse me-4 mb-3" *ngIf="!notificationTab.isRead">
              <button class="btn-mark-as-read" type="button" [ngClass]="getButtonClass(notificationTab)" (click)="markNotificationAsRead(notificationTab)">
                Mark as Read
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </ng-container>
</tm-loading-retry>
