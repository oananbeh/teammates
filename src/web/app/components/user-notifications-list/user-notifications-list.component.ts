import { Component, Input, OnInit } from '@angular/core';
import { finalize } from 'rxjs/operators';
import { NotificationService } from '../../../services/notification.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TableComparatorService } from '../../../services/table-comparator.service';
import { TimezoneService } from '../../../services/timezone.service';
import {
  Notification,
  Notifications,
  NotificationTargetUser,
  ReadNotifications,
} from '../../../types/api-output';
import { SortBy, SortOrder } from '../../../types/sort-properties';
import { ErrorMessageOutput } from '../../error-message-output';
import { collapseAnim } from '../teammates-common/collapse-anim';

export interface NotificationTab {
  notification: Notification;
  hasTabExpanded: boolean;
  isRead: boolean;
  startDate: string;
  endDate: string;
}

/**
 * Component for user notifications list.
 */
@Component({
  selector: 'tm-user-notifications-list',
  templateUrl: './user-notifications-list.component.html',
  styleUrls: ['./user-notifications-list.component.scss'],
  animations: [collapseAnim],
})
export class UserNotificationsListComponent implements OnInit {

  // enum
  NotificationTargetUser: typeof NotificationTargetUser = NotificationTargetUser;
  SortBy: typeof SortBy = SortBy;

  @Input()
  userType: NotificationTargetUser = NotificationTargetUser.GENERAL;

  @Input()
  timezone = '';

  notificationTabs: NotificationTab[] = [];
  notificationsSortBy: SortBy = SortBy.NONE;
  readNotifications: Set<string> = new Set();

  isLoadingNotifications: boolean = false;
  hasLoadingFailed: boolean = false;

  DATE_FORMAT: string = 'DD MMM YYYY';

  constructor(private notificationService: NotificationService,
              private statusMessageService: StatusMessageService,
              private timezoneService: TimezoneService,
              private tableComparatorService: TableComparatorService) { }

  ngOnInit(): void {
    this.loadNotifications();
  }

  loadNotifications(): void {
    this.hasLoadingFailed = false;
    this.isLoadingNotifications = true;

    this.notificationService.getReadNotifications()
      .subscribe({
        next: (readNotifications: ReadNotifications) => {
          this.readNotifications = new Set(readNotifications.readNotifications);
        },
        error: (resp: ErrorMessageOutput) => {
          this.hasLoadingFailed = true;
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      });

    this.notificationService.getAllNotificationsForTargetUser(this.userType)
      .pipe(finalize(() => { this.isLoadingNotifications = false; }))
      .subscribe({
        next: (notifications: Notifications) => {
          notifications.notifications.forEach((notification: Notification) => {
            this.notificationTabs.push({
              notification,
              hasTabExpanded: !this.readNotifications.has(notification.notificationId),
              isRead: this.readNotifications.has(notification.notificationId),
              startDate: this.timezoneService.formatToString(
                  notification.startTimestamp, this.timezone, this.DATE_FORMAT,
              ),
              endDate: this.timezoneService.formatToString(
                  notification.endTimestamp, this.timezone, this.DATE_FORMAT,
              ),
            });
          });
          this.sortNotificationsBy(this.notificationsSortBy);
        },
        error: (resp: ErrorMessageOutput) => {
          this.hasLoadingFailed = true;
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      });
  }

  toggleCard(notificationTab: NotificationTab): void {
    notificationTab.hasTabExpanded = !notificationTab.hasTabExpanded;
  }

  markNotificationAsRead(notificationTab: NotificationTab): void {
    const notification: Notification = notificationTab.notification;
    this.notificationService.markNotificationAsRead({
      notificationId: notification.notificationId,
      endTimestamp: notification.endTimestamp,
    })
      .subscribe({
        next: (readNotifications: ReadNotifications) => {
          this.readNotifications = new Set(readNotifications.readNotifications);
          notificationTab.isRead = true;
          this.statusMessageService.showSuccessToast('Notification marked as read.');
          notificationTab.hasTabExpanded = false;
        },
        error: (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      });
  }

  getBodyTextClass(notificationTab: NotificationTab): string {
    return notificationTab.isRead ? 'card-body' : 'card-body pb-0';
  }

  getButtonClass(notificationTab: NotificationTab): string {
    return `btn btn-${notificationTab.notification.style.toLowerCase()}`;
  }

  /**
   * Checks the option selected to sort notifications.
   */
  isSelectedForSorting(by: SortBy): boolean {
    return this.notificationsSortBy === by;
  }

  /**
   * Sorts the notifications according to the selected option.
   */
  sortNotificationsBy(by: SortBy): void {
    this.notificationsSortBy = by;
    this.notificationTabs.sort(this.sortTabsBy(by));
  }

  /**
   * Sorts the notification tabs in order.
   */
  sortTabsBy(by: SortBy): ((a: NotificationTab, b: NotificationTab) => number) {
    return ((a: NotificationTab, b: NotificationTab): number => {
      let strA: string;
      let strB: string;
      let order: SortOrder;
      switch (by) {
        case SortBy.NOTIFICATION_START_TIME:
          strA = a.notification.startTimestamp.toString();
          strB = b.notification.startTimestamp.toString();
          order = SortOrder.DESC;
          break;
        case SortBy.NOTIFICATION_END_TIME:
          strA = a.notification.endTimestamp.toString();
          strB = b.notification.endTimestamp.toString();
          order = SortOrder.ASC;
          break;
        default:
          strA = '';
          strB = '';
          order = SortOrder.ASC;
      }
      return this.tableComparatorService.compare(by, order, strA, strB);
    });
  }
}
