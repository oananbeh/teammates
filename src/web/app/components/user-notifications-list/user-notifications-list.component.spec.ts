import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import moment from 'moment-timezone';
import { of } from 'rxjs';
import SpyInstance = jest.SpyInstance;
import { NotificationTab, UserNotificationsListComponent } from './user-notifications-list.component';
import { NotificationService } from '../../../services/notification.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { Notification, NotificationStyle, NotificationTargetUser } from '../../../types/api-output';
import { MarkNotificationAsReadRequest } from '../../../types/api-request';
import { SortBy } from '../../../types/sort-properties';
import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
import { PanelChevronModule } from '../panel-chevron/panel-chevron.module';
import { TeammatesCommonModule } from '../teammates-common/teammates-common.module';

describe('UserNotificationsListComponent', () => {
  let component: UserNotificationsListComponent;
  let notificationService: NotificationService;
  let statusMessageService: StatusMessageService;
  let fixture: ComponentFixture<UserNotificationsListComponent>;

  const testNotificationOne: Notification = {
    notificationId: 'notification1',
    startTimestamp: new Date('2017-09-15T09:30+00:00').getTime(),
    endTimestamp: new Date('2050-09-15T09:30+00:00').getTime(),
    createdAt: new Date('2017-09-15T09:30+00:00').getTime(),
    style: NotificationStyle.SUCCESS,
    targetUser: NotificationTargetUser.GENERAL,
    title: 'valid title 1',
    message: 'valid message 1',
    shown: false,
  };

  const testNotificationTwo: Notification = {
    notificationId: 'notification2',
    startTimestamp: new Date('2018-12-15T09:30+00:00').getTime(),
    endTimestamp: new Date('2050-11-15T09:30+00:00').getTime(),
    createdAt: new Date('2018-11-15T09:30+00:00').getTime(),
    style: NotificationStyle.DANGER,
    targetUser: NotificationTargetUser.GENERAL,
    title: 'valid title 2',
    message: 'valid message 2',
    shown: false,
  };

  const timezone: string = 'UTC';
  const DATE_FORMAT: string = 'DD MMM YYYY';

  const getNotificationTabs = (notifications: Notification[], readNotifications: string[] = []): NotificationTab[] => {
    const notificationTabs: NotificationTab[] = [];
    notifications.forEach((notification) => {
      const notificationTab: NotificationTab = {
        notification,
        hasTabExpanded: !readNotifications.includes(notification.notificationId),
        isRead: readNotifications.includes(notification.notificationId),
        startDate: moment(notification.startTimestamp).tz(timezone).format(DATE_FORMAT),
        endDate: moment(notification.endTimestamp).tz(timezone).format(DATE_FORMAT),
      };
      notificationTabs.push(notificationTab);
    });
    return notificationTabs;
  };

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [UserNotificationsListComponent],
      imports: [
        BrowserAnimationsModule,
        HttpClientTestingModule,
        PanelChevronModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
        TeammatesCommonModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserNotificationsListComponent);
    notificationService = TestBed.inject(NotificationService);
    statusMessageService = TestBed.inject(StatusMessageService);
    component = fixture.componentInstance;
    fixture.detectChanges();
    component.timezone = timezone;
    component.isLoadingNotifications = false;
    component.hasLoadingFailed = false;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should load notifications from APIs correctly', () => {
    // two notifications, only first one has been read
    const getNotificationSpy = jest.spyOn(notificationService, 'getAllNotificationsForTargetUser').mockReturnValue(of({
      notifications: [testNotificationOne, testNotificationTwo],
    }));
    const getReadNotificationSpy = jest.spyOn(notificationService, 'getReadNotifications').mockReturnValue(of({
      readNotifications: [
        testNotificationOne.notificationId,
      ],
    }));
    component.loadNotifications();
    expect(getNotificationSpy).toHaveBeenCalledTimes(1);
    expect(getReadNotificationSpy).toHaveBeenCalledTimes(1);
    expect(component.notificationTabs).toEqual(getNotificationTabs(
      [testNotificationOne, testNotificationTwo],
      [testNotificationOne.notificationId],
    ));
  });

  it('should mark notification as read when button is clicked', () => {
    const apiSpy: SpyInstance = jest.spyOn(notificationService, 'markNotificationAsRead')
      .mockImplementation((request: MarkNotificationAsReadRequest) => {
        expect(request.notificationId).toEqual(testNotificationOne.notificationId);
        return of({
          readNotifications: [request.notificationId],
        });
      });
    const messageSpy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
      .mockImplementation((args: string) => {
        expect(args).toEqual('Notification marked as read.');
      });

    component.notificationTabs = getNotificationTabs([testNotificationOne]);
    fixture.detectChanges();
    expect(component.notificationTabs[0].hasTabExpanded).toBeTruthy();
    fixture.debugElement.query(By.css('.btn-mark-as-read')).nativeElement.click();
    expect(apiSpy).toHaveBeenCalledTimes(1);
    expect(messageSpy).toHaveBeenCalledTimes(1);
    // check that it is no longer expanded
    expect(component.notificationTabs[0].hasTabExpanded).toBeFalsy();
  });

  it('should collapse an expanded tab when the header is clicked', () => {
    component.notificationTabs = getNotificationTabs([testNotificationOne]);
    fixture.detectChanges();

    expect(component.notificationTabs[0].hasTabExpanded).toBeTruthy();
    fixture.debugElement.query(By.css('.card-header')).nativeElement.click();
    expect(component.notificationTabs[0].hasTabExpanded).toBeFalsy();
  });

  it('should sort notifications in correct order', () => {
    component.notificationTabs = getNotificationTabs([testNotificationOne, testNotificationTwo]);
    fixture.detectChanges();

    // default order already checked above
    component.sortNotificationsBy(SortBy.NOTIFICATION_START_TIME);
    expect(component.notificationTabs[0].notification.startTimestamp
      >= component.notificationTabs[1].notification.startTimestamp).toBeTruthy();

    component.sortNotificationsBy(SortBy.NOTIFICATION_END_TIME);
    expect(component.notificationTabs[0].notification.endTimestamp
      <= component.notificationTabs[1].notification.endTimestamp).toBeTruthy();
  });

  it('should snap with default fields when loading', () => {
    component.isLoadingNotifications = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when it fails to load', () => {
    component.hasLoadingFailed = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with no notifications', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when it loads the provided notifications', () => {
    component.notificationTabs = getNotificationTabs([testNotificationOne, testNotificationTwo]);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when all loaded notifications are read', () => {
    component.notificationTabs = getNotificationTabs(
      [testNotificationOne, testNotificationTwo],
      [
        testNotificationOne.notificationId,
        testNotificationTwo.notificationId,
      ],
    );
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when it sorts the notification by start time', () => {
    component.notificationTabs = getNotificationTabs([testNotificationTwo, testNotificationOne]);
    component.notificationsSortBy = SortBy.NOTIFICATION_START_TIME;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when it sorts the notification by end time', () => {
    component.notificationTabs = getNotificationTabs([testNotificationOne, testNotificationTwo]);
    component.notificationsSortBy = SortBy.NOTIFICATION_END_TIME;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
