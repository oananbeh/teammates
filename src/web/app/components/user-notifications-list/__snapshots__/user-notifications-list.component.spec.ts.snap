// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserNotificationsListComponent should snap when all loaded notifications are read 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed="false"
  isLoadingNotifications="false"
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy="0"
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><div
    class="row mb-3"
  >
    <div
      class="col-12"
    >
      <div
        class="float-end"
      >
        <strong
          class="d-inline"
        >
           Sort By: 
        </strong>
        <div
          class="btn-group"
          data-toggle="buttons"
        >
          <button
            class="btn btn-light"
            id="sort-end-time"
          >
            End Time
          </button>
          <button
            class="btn btn-light"
            id="sort-start-time"
          >
            Start Time
          </button>
        </div>
      </div>
    </div>
  </div><tm-loading-retry>
    <div>
    </div>
    <div
      id="notification-tabs"
    >
      <div>
        <div
          class="card"
          id="notification1"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-success"
          >
            <strong
              class="text-break"
            >
              valid title 1 [15 Sep 2017 - 15 Sep 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="false"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-down"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="card"
          id="notification2"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-danger"
          >
            <strong
              class="text-break"
            >
              valid title 2 [15 Dec 2018 - 15 Nov 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="false"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-down"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;

exports[`UserNotificationsListComponent should snap when it fails to load 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed={[Function Boolean]}
  isLoadingNotifications="false"
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy="0"
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><tm-loading-retry>
    <div
      class="text-center"
    >
      <p
        class="text-muted m-1"
      >
        Failed to load notifications
      </p>
      <button
        class="btn btn-primary"
      >
        <i
          aria-hidden="true"
          class="fas fa-redo"
        />
        Â Retry
      </button>
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;

exports[`UserNotificationsListComponent should snap when it loads the provided notifications 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed="false"
  isLoadingNotifications="false"
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy="0"
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><div
    class="row mb-3"
  >
    <div
      class="col-12"
    >
      <div
        class="float-end"
      >
        <strong
          class="d-inline"
        >
           Sort By: 
        </strong>
        <div
          class="btn-group"
          data-toggle="buttons"
        >
          <button
            class="btn btn-light"
            id="sort-end-time"
          >
            End Time
          </button>
          <button
            class="btn btn-light"
            id="sort-start-time"
          >
            Start Time
          </button>
        </div>
      </div>
    </div>
  </div><tm-loading-retry>
    <div>
    </div>
    <div
      id="notification-tabs"
    >
      <div>
        <div
          class="card"
          id="notification1"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-success"
          >
            <strong
              class="text-break"
            >
              valid title 1 [15 Sep 2017 - 15 Sep 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="true"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-up"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
          <div
            class="mb-0 ng-trigger ng-trigger-collapseAnim ng-animating"
          >
            <div
              class="card-body pb-0"
            >
              <div
                class="notification-message"
              >
                valid message 1
              </div>
            </div>
            <div
              class="d-flex flex-row-reverse me-4 mb-3"
              style=""
            >
              <button
                class="btn-mark-as-read btn btn-success"
                type="button"
              >
                 Mark as Read 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="card"
          id="notification2"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-danger"
          >
            <strong
              class="text-break"
            >
              valid title 2 [15 Dec 2018 - 15 Nov 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="true"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-up"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
          <div
            class="mb-0 ng-trigger ng-trigger-collapseAnim ng-animating"
          >
            <div
              class="card-body pb-0"
            >
              <div
                class="notification-message"
              >
                valid message 2
              </div>
            </div>
            <div
              class="d-flex flex-row-reverse me-4 mb-3"
              style=""
            >
              <button
                class="btn-mark-as-read btn btn-danger"
                type="button"
              >
                 Mark as Read 
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;

exports[`UserNotificationsListComponent should snap when it sorts the notification by end time 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed="false"
  isLoadingNotifications="false"
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy={[Function Number]}
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><div
    class="row mb-3"
  >
    <div
      class="col-12"
    >
      <div
        class="float-end"
      >
        <strong
          class="d-inline"
        >
           Sort By: 
        </strong>
        <div
          class="btn-group"
          data-toggle="buttons"
        >
          <button
            class="btn btn-light"
            disabled=""
            id="sort-end-time"
          >
            End Time
          </button>
          <button
            class="btn btn-light"
            id="sort-start-time"
          >
            Start Time
          </button>
        </div>
      </div>
    </div>
  </div><tm-loading-retry>
    <div>
    </div>
    <div
      id="notification-tabs"
    >
      <div>
        <div
          class="card"
          id="notification1"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-success"
          >
            <strong
              class="text-break"
            >
              valid title 1 [15 Sep 2017 - 15 Sep 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="true"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-up"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
          <div
            class="mb-0 ng-trigger ng-trigger-collapseAnim ng-animating"
          >
            <div
              class="card-body pb-0"
            >
              <div
                class="notification-message"
              >
                valid message 1
              </div>
            </div>
            <div
              class="d-flex flex-row-reverse me-4 mb-3"
              style=""
            >
              <button
                class="btn-mark-as-read btn btn-success"
                type="button"
              >
                 Mark as Read 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="card"
          id="notification2"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-danger"
          >
            <strong
              class="text-break"
            >
              valid title 2 [15 Dec 2018 - 15 Nov 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="true"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-up"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
          <div
            class="mb-0 ng-trigger ng-trigger-collapseAnim ng-animating"
          >
            <div
              class="card-body pb-0"
            >
              <div
                class="notification-message"
              >
                valid message 2
              </div>
            </div>
            <div
              class="d-flex flex-row-reverse me-4 mb-3"
              style=""
            >
              <button
                class="btn-mark-as-read btn btn-danger"
                type="button"
              >
                 Mark as Read 
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;

exports[`UserNotificationsListComponent should snap when it sorts the notification by start time 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed="false"
  isLoadingNotifications="false"
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy={[Function Number]}
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><div
    class="row mb-3"
  >
    <div
      class="col-12"
    >
      <div
        class="float-end"
      >
        <strong
          class="d-inline"
        >
           Sort By: 
        </strong>
        <div
          class="btn-group"
          data-toggle="buttons"
        >
          <button
            class="btn btn-light"
            id="sort-end-time"
          >
            End Time
          </button>
          <button
            class="btn btn-light"
            disabled=""
            id="sort-start-time"
          >
            Start Time
          </button>
        </div>
      </div>
    </div>
  </div><tm-loading-retry>
    <div>
    </div>
    <div
      id="notification-tabs"
    >
      <div>
        <div
          class="card"
          id="notification2"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-danger"
          >
            <strong
              class="text-break"
            >
              valid title 2 [15 Dec 2018 - 15 Nov 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="true"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-up"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
          <div
            class="mb-0 ng-trigger ng-trigger-collapseAnim ng-animating"
          >
            <div
              class="card-body pb-0"
            >
              <div
                class="notification-message"
              >
                valid message 2
              </div>
            </div>
            <div
              class="d-flex flex-row-reverse me-4 mb-3"
              style=""
            >
              <button
                class="btn-mark-as-read btn btn-danger"
                type="button"
              >
                 Mark as Read 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="card"
          id="notification1"
        >
          <div
            class="card-header cursor-pointer mb-0 alert alert-success"
          >
            <strong
              class="text-break"
            >
              valid title 1 [15 Sep 2017 - 15 Sep 2050]
            </strong>
            <div
              class="card-header-btn-toolbar"
            >
              <tm-panel-chevron>
                <button
                  aria-expanded="true"
                  aria-label="Expand panel"
                  class="chevron btn-course btn btn-primary btn-sm"
                  type="button"
                >
                  <i
                    class="fas fa-chevron-up"
                    style="color: white;"
                  />
                </button>
              </tm-panel-chevron>
            </div>
          </div>
          <div
            class="mb-0 ng-trigger ng-trigger-collapseAnim ng-animating"
          >
            <div
              class="card-body pb-0"
            >
              <div
                class="notification-message"
              >
                valid message 1
              </div>
            </div>
            <div
              class="d-flex flex-row-reverse me-4 mb-3"
              style=""
            >
              <button
                class="btn-mark-as-read btn btn-success"
                type="button"
              >
                 Mark as Read 
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;

exports[`UserNotificationsListComponent should snap with default fields when loading 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed="false"
  isLoadingNotifications={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy="0"
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
    <div
      id="notification-tabs"
    >
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;

exports[`UserNotificationsListComponent should snap with no notifications 1`] = `
<tm-user-notifications-list
  DATE_FORMAT={[Function String]}
  NotificationTargetUser={[Function Object]}
  SortBy={[Function Object]}
  hasLoadingFailed="false"
  isLoadingNotifications="false"
  notificationService={[Function NotificationService]}
  notificationTabs={[Function Array]}
  notificationsSortBy="0"
  readNotifications={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
  timezone={[Function String]}
  timezoneService={[Function TimezoneService]}
  userType={[Function String]}
>
  <div
    id="notifications-timezone"
  >
    <p>
      All dates are displayed in UTC time.
    </p>
  </div><tm-loading-retry>
    <div>
      <div
        class="alert alert-warning margin-top-30px"
        role="alert"
      >
         There are no active notifications. 
      </div>
    </div>
    <div
      id="notification-tabs"
    >
    </div>
  </tm-loading-retry>
</tm-user-notifications-list>
`;
